library/alloc/src/vec/mod.rs

```diff
@@ -2586,11 +2630,10 @@ impl<T, A: Allocator> Vec<T, A> {
                 self.reserve(lower.saturating_add(1));
             }
             unsafe {
-                ptr::write(self.as_mut_ptr().add(len), element);
                 // Since next() executes user code which can panic we have to bump the length
-                // after each step.
-                // NB can't overflow since we would have had to alloc the address space
-                self.set_len(len + 1);
+                // after each step, i.e. we can't move that after the loop.
+                // NB can't fail since we would have had to alloc the address space
+                self.push_within_capacity(element).unwrap_unchecked()
```

Comment:
Is this .unwrap_unchecked() done to avoid a unused result warn-allow annotation?

Reply:
It's to tell the compiler that that the length check (the `Err` case) isn't needed, otherwise I would have used `let _ =` like in the example.

Reply:
In this case you're basically adding an unnecessary check coming from `push_within_capacity` and then tell the compiler to skip it using the `unreachable_unchecked` hint inside `unwrap_unchecked`. It's kinda ok, but it may degrade both compile and sometimes runtime performance as well. 
I think it would be better to have a separate, probably even public `push_unchecked` method that only does write + inc length and then use it where appropriate. 
Also, the length increment could benefit from the `unchecked_add` intrinsic in some niche cases since we're guaranteed that length can't be greater than capacity and capacity can't overflow. Optimization example: https://rust.godbolt.org/z/9fv8q7T9r

Reply:
> It's kinda ok, but it may degrade both compile and sometimes runtime performance as well.

Especially debug builds are more affected by this.

Reply:
Reverted this part since the perf impact tended to be negative.

Reply:
@MSxDOS tried that in #89152, didn't have any impact on the compiler and microbenchmarks didn't show any difference either.